python/comps.py

```diff
@@ -193,103 +244,102 @@ def get_comps(
+        observation_price = observation_matrix[obs_idx, observation_price_idx]
+        bin_found = False
+        for bin_idx in range(num_price_bins):
+            bin_min = price_bin_matrix[bin_idx, price_bin_min_idx]
+            bin_max = price_bin_matrix[bin_idx, price_bin_max_idx]
+            bin_id = price_bin_matrix[bin_idx, price_bin_idx]
+
             if (
                 # Since we expect the price bins to be non-overlapping with
                 # no gaps and an integer difference of 1 between ranges, the
                 # ranges can be treated as inclusive on both ends
-                observation_price >= bin[price_bin_min_idx] and
-                observation_price <= bin[price_bin_max_idx]
+                observation_price >= bin_min and observation_price <= bin_max
             ):
-                output_matrix[obs_idx] = bin[price_bin_idx]
+                output_vector[obs_idx, 0] = bin_id
+                bin_found = True
                 break
-        else:
-            raise ValueError(
-                f"Observation {obs_idx} did not match any price bins"
-            )
 
-    return output_matrix
+        if not bin_found:
+            # Set a special value to indicate an error, since taichi doesn't
+            # support runtime errors
+            output_vector[obs_idx, 0] = -1
 
 
-@nb.njit(fastmath=True, parallel=True)
+@ti.kernel
 def _get_top_n_comps(
-    leaf_node_matrix, comparison_leaf_node_matrix, weights_matrix, num_comps
+    leaf_node_matrix: ti.types.ndarray(),
+    comparison_leaf_node_matrix: ti.types.ndarray(),
+    weights_matrix: ti.types.ndarray(),
+    all_top_n_idxs: ti.types.ndarray(),
+    all_top_n_scores: ti.types.ndarray(),
+    num_comps: int,
+    num_observations: int,
+    num_possible_comparisons: int,
+    num_trees: int
 ):
     """Helper function that takes matrices of leaf node assignments for
     observations in a tree model, a matrix of weights for each obs/tree, and an
     integer `num_comps`, and returns a matrix where each observation is scored
     by similarity to observations in the comparison matrix and the top N scores
     are returned along with the indexes of the comparison observations."""
-    num_observations = len(leaf_node_matrix)
-    num_possible_comparisons = len(comparison_leaf_node_matrix)
-    idx_dtype = np.int32
-    score_dtype = np.float32
-
-    # Store scores and indexes in two separate arrays rather than a 3d matrix
-    # for simplicity (array of tuples does not convert to pandas properly).
-    # Indexes default to -1, which is an impossible index and so is a signal
-    # that no comp was found
-    all_top_n_idxs = np.full((num_observations, num_comps), -1, dtype=idx_dtype)
-    all_top_n_scores = np.zeros((num_observations, num_comps), dtype=score_dtype)
-
-    for x_i in nb.prange(num_observations):
-        top_n_idxs = np.full(num_comps, -1, dtype=idx_dtype)
-        top_n_scores = np.zeros(num_comps, dtype=score_dtype)
```

Comment:
Arrays cannot be defined inside a taichi kernel, and must instead be defined outside its scope and then passed in as an argument; hence, we needed to refactor this function to remove the dynamic allocation of arrays inside the context of the function.